### DISCUSSION TITLE
Font-x-size property for CSS

### DISCUSSION THREAD

**Discussion by Myndex** [w3c/low-vision-a11y-tf#118] [Ideas]
I just posted the proposal I discussed previously regarding a needed property to set x-height directly in CSS.

https://github.com/w3c/csswg-drafts/issues/6709

A

**Comment by mraccess77** [comment]
Hi Andy, it might be useful to followup on that discussion with a reference to why font-size-adjust is not sufficient even though you spelled that out already.  

At this point - this CSS related item is something that the LVTF is not currently tackling with our current workload.  Some of us support the idea and understand the need.   My recommendation is that our group could add some support to your proposal but that we would not specifically actively work as a group to shepherd the proposal.  I'd recommend that you continue to work with the CSS working group and others like APA to continue your efforts.

As WCAG 3.0 progresses - if the AGWG determines this is needed to proceed for new criteria then this may be something the AGWG would need to also lend support for.

**Reply by Myndex** [reply]
Hi @mraccess77  Jonathan,

> _Hi Andy, it might be useful to followup on that discussion with a reference to why font-size-adjust is not sufficient even though you spelled that out already._

I did link to the examples page, which has further details... I suspect this is going to be a conversation — ` font-size-adjust: ` is a much maligned property, all bus discarded by browsers, now down to only Firefox... On the examples page I discuss also why it is not sufficient. Nevertheless, I expect some further discussion here.

> _At this point - this CSS related item is something that the LVTF is not currently tackling with our current workload. Some of us support the idea and understand the need._

**Just FYI,** When I reached out to leadership in January about getting a vision subgroup going again, the PRINCIPAL reason I identified was the need for liaisons into CSS. AFAIC this is the most pressing issue, because the "train is leaving the station" and there seems to be no one promoting these very basic, fundamental issues over there.

> _My recommendation is that our group could add some support to your proposal but that we would not specifically actively work as a group to shepherd the proposal. I'd recommend that you continue to work with the CSS working group and others like APA to continue your efforts._

I will continue, and possibly should become more involved with APA. If I understand how the process works, I assume this is going to go to a vote at some point, and it would be good to have a united front on these kinds of issues. This is important, so it is in my higher priority list, because this is needed for what I am working on for WCAG 3. When we specify a standard such as "font size" it needs to be meaningful and translate to something concrete and consistent. 

### But at the moment the best we can do is to say _"compare to this reference font because we have literally no other way to specify anything that relates to the real world."_

And this goes for font WEIGHT too. And also letter and line spacing. The single most useful "size" when it comes to readability is the x-height. Even then, there are variations — and weight is the next significant issue, though that will take a technology to address (for it to be automated at least). Without any standardized form of setting size and weight, there is literally no actual path for useful automated testability. 

Before we can move forward with useful standards on text for advanced things like assessing weight, we need to get this one most basic thing fixed, and that is setting size standards per x-height. And then linking ` em ` and ` line-height: ` and the related to that x-height size for proper cascading/inheritance, etc.

**_Side note on the importance relative to contrast:_**
Font size and weight are a _bigger issue for contrast than colors_ when the spatial frequency is high, such as with body text. When the spatial frequency is low, like for a button or a very big bold headline, then colors are the dominant aspect of contrast. But without affirmative control of size (and weight) as it renders to screen, it is impossible to predict the actual perceived and readable contrast. For small fonts as are used in body text (fonts under 24px) luminance contrast needs to increase to near the max rapidly. 

> As WCAG 3.0 progresses - if the AGWG determines this is needed to proceed for new criteria then this may be something the AGWG would need to also lend support for.

I'm really not the guy for "internal organizational politics" though. But if I understand how this organization works, there will be some voting, and it would be good to have something of a united front if at all possible.

There is a list of needed items, and many of them rely first on getting font sizing fixed. Some of the others:

- Consistent font weight assessment, preferably automated. PANOSE never really lived up to potential.
- "Adjustment" properties so a user style sheet could make adjustments to a given page without overwriting.
    - The name "font-size-adjust" seems much better suited for this task than for setting x-height.
    - Ideally, a user style sheet could adjust without overwriting: size, weight, tracking/leading, lightness/darkness, and hue (in degrees)

> _Could we say something in WCAG like the lowercase letter height is at least 1/2 of the font size? Seems like we could also have issues if the ascenders were also too short compared to the total height - so it seems like there is some range. And also no more than 3/4 of the total font size?_

In the FPWD, I had indicated that the sizes were based on a font where the x-height was at least 0.5 the font size, and to compare fonts chosen to the short list of reference fonts.

As to the range of ½ to ¾ of the font size: that's a 50% difference, it means that for a font set at 16px, the lowercase could be between 8px and 12px. From the contrast point of view, a, 8px to 12px size difference is like the difference between 4.5:1 and 7:1 in WCAG 2, or in APCA, a difference of L<sup>c</sup> 25, i.e.  L<sup>c</sup> 55 to L<sup>c</sup> 80. That's a massive variance, and it is why being able to affirmatively set size using x-height directly is so very important.

For the record, "good" screen fonts have an x-height ratio of about 0.6:1. A bad screen font is like _Times New Roman_ which has an x-height ratio of 0.45:1 — less than ½ !!

_Times New Roman_ is one standard font that demonstrates the need here, just as the atrocious _"Courier New"_ demonstrates the need for effective weight assessment.

<img src="https://user-images.githubusercontent.com/42009457/136056260-28aedcd3-6f0f-4037-bc2e-446d9aa0a1bd.png" width="600"/>


Thank you,


Andy

**Comment by mraccess77** [comment]
Could we say something in WCAG like the lowercase letter height is at least 1/2 of the font size?  Seems like we could also have issues if the ascenders were also too short compared to the total height - so it seems like there is some range.  And also no more than 3/4 of the total font size?
